python/ray/tune/checkpoint_manager.py

```diff
@@ -0,0 +1,139 @@
+        if self.priority == other.priority:
+            return 0
+        return -1 if self.priority < other.priority else 1
+
+    def __lt__(self, other):
+        return self.priority < other.priority
+
+
+class CheckpointManager(object):
+    """Manages checkpoints on the driver for a trial."""
+
+    def __init__(self, keep_checkpoints_num, checkpoint_score_attr):
+        """Initializes a new CheckpointManager.
+
+        Args:
+            keep_checkpoints_num (int): Keep at least this many checkpoints.
+            checkpoint_score_attr (str): Attribute to use to determine which
+                checkpoints to keep.
+        """
+        self.keep_checkpoints_num = keep_checkpoints_num or float("inf")
+        assert self.keep_checkpoints_num > 0, (
+            "keep_checkpoints_num must be greater than 0.")
+        self._checkpoint_score_desc = checkpoint_score_attr.startswith("min-")
+        if self._checkpoint_score_desc:
+            self._checkpoint_score_attr = checkpoint_score_attr[4:]
+        else:
+            self._checkpoint_score_attr = checkpoint_score_attr
+
+        self.newest_checkpoint = Checkpoint(Checkpoint.MEMORY, None)
+        self._best_checkpoints = []
+        self._membership = set()
+
+    def on_checkpoint(self, checkpoint):
+        """Starts tracking checkpoint metadata on checkpoint.
+
+        Sets newest checkpoint. Deletes previous checkpoint as long as it isn't
+        one of the best ones. Also deletes the worst checkpoint if at capacity.
+
+        Args:
+            checkpoint (Checkpoint): Trial state checkpoint.
+
+        Raises:
+            KeyError if checkpoint_score_attr not in last_result of checkpoint.
+        """
+        old_checkpoint = self.newest_checkpoint
+        self.newest_checkpoint = checkpoint
+
+        try:
+            queue_item = QueueItem(self._priority(checkpoint), checkpoint)
+        except KeyError:
+            if old_checkpoint not in self._membership:
+                old_checkpoint.delete()
+            raise ValueError(
+                "Result dict has no key: {}. checkpoint_score_attr must be "
+                "set to a key in the result dict.".format(
+                    self._checkpoint_score_attr))
+
+        if len(self._best_checkpoints) < self.keep_checkpoints_num:
+            heapq.heappush(self._best_checkpoints, queue_item)
+            self._membership.add(checkpoint)
+        elif queue_item.priority >= self._best_checkpoints[0].priority:
+            worst = heapq.heappushpop(self._best_checkpoints, queue_item).value
+            self._membership.add(checkpoint)
+            if worst in self._membership:
+                self._membership.remove(worst)
+            worst.delete()
+
+        # Remove the old checkpoint if it isn't one of the best ones.
+        if old_checkpoint not in self._membership:
+            old_checkpoint.delete()
+
+    def best_checkpoints(self):
+        """Returns best checkpoints, sorted by priority."""
```

Comment:
BTW, the concept of 'priority' is not a Tune concept - maybe instead say "score"?